src/wayland/fifo/mod.rs

```diff
@@ -0,0 +1,406 @@
+                if is_managed && is_initial {
+                    add_pre_commit_hook::<D, _>(&surface, |_, _, surface| {
+                        let fifo_barrier = with_states(surface, |states| {
+                            let fifo_state = *states.cached_state.get::<FifoCachedState>().pending();
+
+                            // The pending state will contain any previously set barrier on this surface
+                            // In case this commit updates the barrier with `set_barrier`, but also mandates to
+                            // wait for a previously set barrier it is important to first retrieve the previously
+                            // set barrier to not overwrite it with our own.
+                            let fifo_barrier = fifo_state
+                                .wait_barrier
+                                .then(|| {
+                                    states
+                                        .cached_state
+                                        .get::<FifoBarrierCachedState>()
+                                        .pending()
+                                        .barrier
+                                        .take()
+                                })
+                                .flatten();
+
+                            // If requested set the barrier for this commit.
+                            // The barrier will be available for the next commit requesting to wait on it
+                            // in the pending state.
+                            // The barrier will also be either put in the current state in case this commit
+                            // is not blocked or into a transaction otherwise eventually ending in the current
+                            // state when it is unblocked.
+                            if fifo_state.set_barrier {
+                                states
+                                    .cached_state
+                                    .get::<FifoBarrierCachedState>()
+                                    .pending()
+                                    .barrier = Some(Barrier::new(false));
+                            }
+
+                            fifo_barrier
+                        });
+
+                        if let Some(barrier) = fifo_barrier {
+                            // If multiple consecutive commits only call wait_barrier, but not set_barrier
+                            // we might end up with the same barrier in multiple commits. It could happen
+                            // that the barrier is already signaled in which case there is no need to
+                            // further delay this commit
+                            //
+                            // In addition the spec also defines that the constraint must be ignored for
+                            // sync subsurfaces
+                            if !barrier.is_signaled() || is_sync_subsurface(surface) {
```

Comment:
What if the surface is set sync afterwards? Given the docs state that `wait_barrier` should be the even ignored, I feel like there are cases, where we need to clear the blocker later (or the blocker needs to check the surface's sync state upon evaluation somehow).

Reply:
Are there any guarantees for a surface being set sync afterwards? This is already racy imo as the blocker could already be cleared and the state applied. So not sure we really need to handle that. Checking in the blocker won't help as we only evaluate it on `blocker_cleared`. I would rather not add some kind of callback logic for surfaces being set sync. We theoretically could check it in either `pre_paint` or `post_paint`.

But I will ask in #wayland for clarification

Reply:
okay, so I was actually redirected to this issue :) https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/362 which also references this one https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/361

while the latter has been closed and the former is not merged yet we can conclude that commit time is the correct interpretation (unless I misunderstood something in the discussion of 362, got a bit lost tbh in between)

Reply:
nice, I actually got the check backwards. should be fixed now and skip the blocker if the barrier is already signaled or the surface is sync